R/CageData_Integration.R

```diff
@@ -0,0 +1,201 @@
+  ends <- ifelse(x[, 6]=="-",x[, 3], x[, 3]-1)
+  gr <- GRanges(x[, 1],IRanges(starts,ends))
+  strand(gr) <-x[, 6]
+  score(gr) <-x[, 5]
+  if(ncol(x) > 6)  mcols(gr) <- x[, 7:ncol(x)]
+  return(gr)
+}
+
+
+#' Get CageData as granges
+#' @param dataName The location of the cage-file
+#' @param filterValue The number of counts(score) to filter on for a tss to pass as hit
+#' @importFrom data.table fread
+getFilteredCageData <- function(dataName, filterValue = 1){
+  if ( .Platform$OS.type == "unix") {
+    if ( file.exists(dataName)) {
+      if( gsub(pattern = ".*\\.","",dataName) == "gzip"){
+      rawCageData <- bedToGR(as.data.frame(fread(paste("gunzip -c",dataName), sep = "\t")))
+      }else if( gsub(pattern = ".*\\.","",dataName) == "bed"){
+        rawCageData <- bedToGR(as.data.frame(fread(dataName, sep = "\t")))
+      }else{ stop("Only bed and gzip formats are supported for filePath")}
+    }else{ stop("Filepath specified does not name existing file.") }
+  }else{ stop("Only unix operating-systems currently support filePath,use cage as GRanges argument instead.") }
+  if( is.null(rawCageData$score)) stop("Found no score column in the cageData-file, bed standard is column 5")
+  message("Loaded cage-file successfully")
+  filteredrawCageData <- rawCageData[rawCageData$score > filterValue,] #filter on score
+
+  #chromosome nameing, need to find a smart way for this, it wont work for all
+  seqlevels(filteredrawCageData) <- sub("chrY","Y",seqlevels(filteredrawCageData))
+  seqlevels(filteredrawCageData) <- sub("chrX","X",seqlevels(filteredrawCageData))
+  seqlevels(filteredrawCageData) <- sub("chrM","MT",seqlevels(filteredrawCageData))
+  return(filteredrawCageData)
+}
+
+#' When reassigning Transcript start site, often you want to add downstream too.
+#' This is a simple way to do that
+#' @param fiveUTRs The 5' leader sequences as GRangesList
+#' @param cds If you want to extend 5' leaders downstream, to catch uorfs going into cds, include it.
+addFirstCdsOnLeaderEnds <- function(fiveUTRs, cds){
+
+  if( length(cds) == 0){
+    warning("cds is empty, returning without using it.")
+    return(fiveUTRs)
+  }
+  if( is.null(names(cds))){
+    warning("cds have no names, returning without using it.")
+    return(fiveUTRs)
+  }
+  matchingNames = names(fiveUTRs) %in% names(cds)
+  if( sum(matchingNames) - length(names(fiveUTRs)) != 0){ # <- check valid names
+    warning("not all cds names matches fiveUTRs names, returning without using cds.")
+    return(fiveUTRs)
+  }
+
+  cdsForUTRs <- cds[names(fiveUTRs)] # get only the ones we need
+
+  firstExons <- phead(cdsForUTRs, 1L) #select first in every, they must be sorted!
+  gr <- unlist(firstExons,use.names = FALSE)
+  gr$cds_id <- NULL # <- remove cds metacols
+  gr$cds_name <- NULL
+  gr$exon_rank <- NULL
+  gr$exon_id <- NA
+  gr$exon_name <- NA
+  gr$exon_rank <- NA
+
+  grl <- relist(gr,firstExons)
+  fiveUTRsWithCdsExons <- pc(fiveUTRs, grl)
```

Comment:
this things should be sorted but check this, also spaces, also droping metacolumns can be done in one line, also subset on the fiveUTRs